crates/fuzzing/src/oracles.rs

```diff
@@ -727,13 +722,135 @@ pub fn differential_wasmi_execution(wasm: &[u8], config: &crate::generators::Con
     }
 
     if &wasmi_buf[..] != &wasmtime_slice[..] {
-        show_wat();
+        show_wat(wasm);
         panic!("memory contents are not equal");
     }
 
     Some(())
 }
 
+/// Perform differential execution between Wasmtime and the official WebAssembly
+/// specification interpreter.
+///
+/// May return `None` if we early-out due to a rejected fuzz config.
+pub fn differential_spec_execution(wasm: &[u8], config: &crate::generators::Config) -> Option<()> {
+    crate::init_fuzzing();
+    debug!("WAT: {}", wasm2wat(wasm));
+    debug!("config: {:#?}", config);
+
+    let wasmtime_vals = run_in_wasmtime(wasm, config, &[]);
+    debug!("Wasmtime returned: {:?}", wasmtime_vals);
+    let spec_vals = wasm_spec_interpreter::interpret(wasm, vec![]);
+    debug!("spec interpreter returned: {:?}", &spec_vals);
+
+    // Match a spec interpreter value against a Wasmtime value. Eventually this
+    // should support references and `v128` (TODO).
+    fn matches(spec_val: &wasm_spec_interpreter::Value, wasmtime_val: &wasmtime::Val) -> bool {
+        match (spec_val, wasmtime_val) {
+            (wasm_spec_interpreter::Value::I32(a), wasmtime::Val::I32(b)) => a == b,
+            (wasm_spec_interpreter::Value::I64(a), wasmtime::Val::I64(b)) => a == b,
+            (wasm_spec_interpreter::Value::F32(a), wasmtime::Val::F32(b)) => *a as u32 == *b,
+            (wasm_spec_interpreter::Value::F64(a), wasmtime::Val::F64(b)) => *a as u64 == *b,
```

Comment:
I think you will want to check for NaNs here, since they may have different bit patterns but be equivalent for our testing concerns (despite not be "equal" under IEEE754...). Even with cranelift's NaN canonicalization enabled, the spec interpreter might not be canonicalizing or might canonicalize to a different bit pattern.

In fact, we [already do this dance](https://github.com/bytecodealliance/wasmtime/blob/main/crates/fuzzing/src/oracles.rs#L329-L356) elsewhere in this crate, so you can factor out an existing method for comparing values.